### PR TITLE
Add python pickle machine (pypm) dissassembler ##anal

### PR DESCRIPTION
**Pull request by swoops**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**
This pull allows r2 to disassemble python pickle opcodes:

```
>  python3 -c 'import pickle;import sys;sys.stdout.buffer.write(pickle.dumps(10))' > /tmp/junk.pickle
> r2 -a pypm /tmp/junk.pickle 
[0x00000000]> pid 3
0x00000000                 8004  PROTO 0x4
0x00000002                 4b0a  BININT1 0xa
0x00000004                   2e  STOP
```
All opcodes have handlers but some are completely untested or validated.  The ops with `// TODO: validate` commented have not been tested at all. Figured I would go ahead and share in case anyone wants to play/break it.

### CODE REVIEW

**Comment by trufae** [comment]
I think it would look better if named just "pickle", without the py_. What do you think?

**Comment by swoops** [comment]
> I think it would look better if named just "pickle", without the py_. What do you think?

Sounds good to me. I will rename the files. 

Is `pypm`, for "python pickle machine" good for the arch name? Like `r2 -pypm file.pickle`? I feel like the `py` helps put it close to `pyc` arch. Also, I don't feel like python pickles are cool enough to deserve a <4 letter arch name. Any thoughts?

Also I tested most of the opcodes and they seem to work. I still need to add tests but otherwise this code is ready I think.

**Comment by swoops** [comment]
I think this is ready for merging now.

There are two places where this comes up short though. Both are in regards to the older opcodes that take in newline terminated strings.

1. A newline terminated argument might be really big, if it's bigger then 128 bytes then the code gives up looking for the newline and considered the op invalid.

2. Some old numerical ops, like `LONG` accept newline terminated strings that are to be parsed into num types, like `LONG "42L"` encoded as `L42L\n` corresponds to python2's `long(42)`. The disassembler here will properly disassemble this opcode, but it won't set the `RAnalOp->val` to `42`, it won't even check if the arg is a number. It just treats it like a string.

**Comment by trufae** [comment]
Pypm sounds like python package manager to me. Why not just pickle? 😃

**Comment by trufae** [comment]
This pr breaks the windows build. Please check the logs https://github.com/radareorg/radare2/runs/7400860775?check_suite_focus=true

**Comment by trufae** [comment]
ping

**Comment by swoops** [comment]
Hopefully windows likes this better. I have very little experience compiling on windows and for good reason :)

File: libr/anal/p/anal_pickle.c [hunk 1]
```diff
@@ -111,7 +111,10 @@ static inline int handle_float(RAnalOp *op, const char *name, int sz, const ut8
 
 static inline char *get_line(const ut8 *buf, int len) {
 	// TODO: find end of large strings through RAnal->iob
-	char *str = strndup ((char *)buf, len);
+	char *str = (char *)malloc (len + 1);
```
Comment (trufae) [inline libr/anal/p/anal_pickle.c:114]: Just use r_str_ndup

File: libr/anal/p/anal_pickle.c [hunk 2]
```diff
@@ -111,7 +111,10 @@ static inline int handle_float(RAnalOp *op, const char *name, int sz, const ut8
...
+	strncpy (str, (char *)buf, len);
+	str[len + 1] = '\0';
```
Comment (trufae) [inline libr/anal/p/anal_pickle.c:116]: this is an overflow. just use r_str_ndup

File: libr/anal/p/anal_pickle.c (hunk 1 above)
Comment (swoops) [reply libr/anal/p/anal_pickle.c]: I knew that existed somewhere but couldn't find it, thanks. Fixed now.

**Comment by trufae** [comment]
should be good after the cast :)

**Pull request merged**